lib/ecto/query/builder.ex

```diff
@@ -419,12 +457,34 @@ defmodule Ecto.Query.Builder do
   defp split_fragment(<<first :: utf8, rest :: binary>>, consumed),
     do: split_fragment(rest, consumed <> <<first :: utf8>>)
 
-  defp escape_call({name, _, args}, type, params, vars, env) do
-    {args, params} = Enum.map_reduce(args, params, &escape(&1, type, &2, vars, env))
+  defp escape_window([], _type, params_acc, _vars, _env),
+    do: {nil, params_acc}
+  defp escape_window([window_name], _type, params_acc, _vars, _env) when is_atom(window_name),
+    do: {window_name, params_acc}
+  defp escape_window([partition_by], type, params_acc, vars, env),
+    do: escape(partition_by, type, params_acc, vars, env)
+
+  defp escape_call({_, _, args} = call, type, params, vars, env) when is_atom(type) do
+    arity = length(args)
+    escape_call(call, List.duplicate(type, arity), params, vars, env)
+  end
+
+  defp escape_call({name, _, args}, types, params, vars, env) when is_list(types) and length(types) >= length(args) do
```

Comment:
I have also added support of function with different arity. If you approve it, we can replace tuples [here](https://github.com/Anber/ecto/blob/7c9096c4a3622977b96c78fabf964e590fb85435/lib/ecto/query/builder.ex#L526)  to lists like `[out_type | in_types]`.

Reply:
@Anber I don't think we should generalize the `length(types) >= length(args)` as we are getting a behaviour specific to two functions and generalizing it to the whole codebase. This is bound to cause issues further down the road. I think you should move the [zip logic to before you call escape_call](https://github.com/elixir-ecto/ecto/pull/2618/files#diff-c04711315c01e29757522ea2764fd871R287).

Reply:
In fact, I would prefer if for now we only added `row_number` and the `*_rank` functions, treat them as aggregators and keep this simple for now. Otherwise there is a chance for this PR to only continue growing in size, making it harder to review and merge. :)

Reply:
Actually, it doesn't matter how many window functions we will implement, because even `row_number` has zero arity and it can't be implemented by using current `call_type`.

Reply:
@Anber I see. So maybe we should do this:

1. Change call_type to return `{[in], out}`

2. Have a specific annotation that tells all functions with variadic arity. For example `@varargs [:lag, :lead, :row_number]`

What do you think?

Reply:
To be more precise, I am just worried about always enabling this behaviour. But if we put it behind a `@varargs` flag, my concerns are addressed and I will be happy. :)

Reply:
Modification of `call_type` definitely looks better than my solution, but I didn't want to touch it without your blessing :)
Maybe I can just modify my `@over_agg` and use it in `call_type`? 
```elixir
  defp call_type(agg, 1) when agg in @over_agg_names, do: Map.get(@over_agg, agg)
```

Reply:
@Anber that's a linear cost. We should implement the proper `call_type/2` clauses and not rely on the map.

However, given the previous discussion about over+window functions, maybe we shouldn't use `call_type` at all and keep them completely distinct, as they seem to have a bunch of behaviour that is specific to them.

Reply:
> that's a linear cost

And what if we unfold call_type on compile-time?
```elixir
  for {agg, types} <- @over_agg do
    defp call_type(unquote(agg), 1), do: types
  end
```

Reply:
That's fine but given the solution above, I think we won't be overriding call_type anyway, will we?

Reply:
Looks like we won't.
Thank you for your patience!